### ISSUE TITLE
gnome: can't edit user profile

### ISSUE THREAD

**Issue by fazo96** [NixOS/nixpkgs#10025]
Changing user Full Name and Profile Picture using `Settings -> Users -> My Profile` from the settings application never gets applied and no errors are displayed.

**Comment by lethalman** [comment]
This is not supported in nixos, and probably will never be. Changing full name will change /etc/passwd, which is managed by nixos. Same applies to datetime settings, it's not possible to change them from the desktop, the same reason why nixos disabled systemd-timedated.
Also changing users from the control panel will never be supported.

Everything that changes system stuff imperatively through systemd services is unsupported.

**Comment by lethalman** [comment]
But, maybe we can disable changing these settings by patching gnome.

**Comment by fazo96** [comment]
Yeah, if possible we should try providing a workaround for chaning the profiles and/or a warning or error message, just to know what's going on (the wiki is a good place for this)

**Comment by vcunat** [comment]
Maybe a big fat warning in the NixOS installation guide that any system-wide settings can only be set by the *NixOS way*.

**Comment by Profpatsch** [comment]
(triage) where should that documentation go?

**Comment by jtojnar** [comment]
I can set profile picture using the Control Center just fine on NixOS with GNOME 3.24, modulo #29983.

Regarding the real name setting, the problem is Control Center does not check if the real name was changed. It should display an error message on failure. Unfortunately, `libaccountsservice` does not return errors on failure: 
https://lazka.github.io/pgi-docs/AccountsService-1.0/classes/User.html#AccountsService.User.set_real_name

We need to convince libaccountsservice maintainers to add functions with error handling and then convince gnome-control-center to use those methods.

**Comment by Profpatsch** [comment]
triage: anything else that should be done?

**Comment by jtojnar** [comment]
We are waiting for AccountService rewrite so that Control Center could display errors.

**Comment by jtojnar** [comment]
Even if it were best-effort, I do not think manipulation file on the computer and then rebuilding is a good idea, there are just too many things that cannot be handled IMO:

- The configuration can change but not yet be rebuilt, changing accounts would apply the other changes as well.
- `/etc/nixos` can be read-only network mount.
- The user might want to track the changes in version control.
- The config is more than just setting nested attributes.

The only way to make configuration.nix machine-editable is if strong limitations on the syntax are enforced by something like `visudo`.

**Comment by Ekleog** [comment]
> Even if it were best-effort, I do not think manipulation file on the computer and then rebuilding is a good idea, there are just too many things that cannot be handled IMO:
>
> - The configuration can change but not yet be rebuilt, changing
> accounts would apply the other changes as well.

This is indeed a problem nixos has and emacs doesn't: emacs just adds
the lines to the file and then changes the running configuration live,
and next boot will take into account all the changes in the file, which
is expected while an automated rebuild isn't expected in nixos.

> - `/etc/nixos` can be read-only network mount.
> - The user might want to track the changes in version control.
> - The config is more than just setting nested attributes.

All these are also potential issues for emacs. What they mean (for
points 2 and 4) is that there must stay a way to make the edit fail --
but that should be handled by the accountservice bugfix, when it will be done.

> The only way to make configuration.nix machine-editable is if strong
> limitations on the syntax are enforced by something like `visudo`.

Not necessarily, emacs handles it by having a zone in the file where it
has its own syntax (and only a zone of it). In nixos terms, this could
translate into a `/etc/nixos/autogenerated.nix` that would be imported
from the default-generated config, and would have the strong limitations
on syntax (and nixos assertions would make sure no user settings
conflict with auto-generated settings).

See also https://discourse.nixos.org/t/user-friendly-nixos-distro/1348 ,
that is related to what I am speaking about -- here the “graphical
configurator” would be the gnome UI.

**Comment by austinbutler** [comment]
It's not clear to me whether I currently _should_ be able to set a custom profile picture through Gnome's built-in settings or not. What I can say is that on unstable, with Gnome 3.36.1, it does not work for me.

```
Apr 15 23:29:12 laptop accounts-daemon[18839]: g_file_info_get_attribute_uint32: assertion 'G_IS_FILE_INFO (info)' failed
Apr 15 23:29:12 laptop accounts-daemon[18839]: g_file_info_get_file_type: assertion 'G_IS_FILE_INFO (info)' failed
Apr 15 23:29:12 laptop accounts-daemon[18839]: g_file_info_get_attribute_uint64: assertion 'G_IS_FILE_INFO (info)' failed
Apr 15 23:29:12 laptop .gnome-control-[23124]: SetIconFile call failed: GDBus.Error:org.freedesktop.Accounts.Error.Failed: file '/tmp/gnome-control-center-user-icon-K0IWI0' is not a regular file
```

**Comment by jtojnar** [comment]
@austinbutler Thanks, completely forgot about this one. Opened https://github.com/NixOS/nixpkgs/issues/85357

**Comment by tholzschuh** [comment]
> I can set profile picture using the Control Center just fine on NixOS with GNOME 3.24, modulo #29983.

This doesn't seem to be the case on GNOME 40.1 for me (in the Wayland variant at least, I haven't tried the XOrg one).

**Comment by notfed** [comment]
> > I can set profile picture using the Control Center just fine on NixOS with GNOME 3.24, modulo #29983.
> 
> This doesn't seem to be the case on GNOME 40.1 for me (in the Wayland variant at least, I haven't tried the XOrg one).

I second this. Can't set a profile picture using Control Center.  It allows me to choose an image, even crop it, but when I click "Select", the crop screen just closes and the selection icon has not changed.

I can however do `cp icon.png /var/lib/AccountsService/icons/$USER` which works fine.

```
$ nixos-version
21.05.1590.63ee5cd99a2 (Okapi)
$ gnome-shell --version
GNOME Shell 40.1
```

**Comment by imsofi** [comment]
The file `/var/lib/AccountsService/icons/$USER` shows that it will look for the profile picture in `/home/$USER/.face`. Symlinking or putting an image file at this location will make it show up on your lock-screen/account.

**Comment by Ashvith10** [comment]
Is this issue still not fixed? Can anyone verify this?

**Comment by pancaek** [comment]
I'm on NixOS 24.05 (stable) and don't have this issue. I'm not using Wayland, though. Using Settings/Gnome Control Center for this works just fine.

**Comment by ghost** [comment]
The original issue cannot be reproduced on my end in GNOME 48 with NixOS 25.05.

**Comment by ghost** [comment]
Changing the profile picture seems to work but changing name doesn't

**Comment by ghost** [comment]
The changed name resets back to the username of the user after a reboot but the image is preserved.
